### PR TITLE
Add ability to connect to IIASA API by application alias

### PR DESCRIPTION
**Pull request by zikolach**
Before this PR review ensure:
- [x] application configs updated accordingly to have necessary aliases (e.g. iamc15)
- [x] deploy relevant auth server changes (exposing alias config parameter in applications list)

### CODE REVIEW

File: pyam/iiasa.py [hunk 1]
```diff
@@ -94,22 +95,36 @@ def __init__(self, name=None, creds=None, base_url=_BASE_URL):
...
-        return valid
+        conn_map = {}
+        for x in r.json():
+            if 'config' in x:
+                env = next((r['value'] for r in x['config']
+                            if r['path'] == 'env'), None)
+                name = x['name']
+                if env is not None:
+                    if env in conn_map:
+                        logger.warning(f'Duplicate instance alias [{env}] '
```
[the remark this answers is not in the document]
Comment (zikolach) [reply pyam/iiasa.py]: updated stickler config

**Comment by danielhuppmann** [comment]
Thanks for the PR @zikolach - but I'm wondering whether keeping the mapping in Python makes sense... Wouldn't this be smarter to be living inside the RestAPI or auth server?

So that the IIASA server only returns the list of "nice-names" and automatically translates a connection established to a nice-name to the correct db?

**Comment by zikolach** [comment]
> this be smarter to be living inside the RestAPI

The mapping is kept on the server already. "Nice name" is a property of application config (property name `env`) and returned as part of application list response (together with app name, product name, UI URL).

**Comment by danielhuppmann** [comment]
Of course the mapping is kept on the server. I just don't see why the best solution for this is to send all details to the client, create a mapping there, do the translation and checking on the client, then open the connection.

Wouldn't it be more elegant and easier for maintenance that `pyam.Connection.connect()` takes any string and sends it to the server and let the server figure out whether its the app name or the nice name...?

**Comment by zikolach** [comment]
> Wouldn't it be more elegant and easier for maintenance

No, it wouldn't (from my point of view). What is the problem you see (and want to solve) with the current solution?

**Comment by danielhuppmann** [comment]
Say a user wants to connect to a particular db. When calling `pyam.iiasa.Connection(<name>)`, in the background on the client side, a mapping of (potentially hundreds) of nice-names to app-names is built up, plus a list of nice-names, irrespective whether this is needed (if `<name>` is the app-name). This seems like excessive back-and-forth of data, when all that is needed is

```
url = '/'.join([self._base_url, 'applications', name, 'config'])
headers = {'Authorization': 'Bearer {}'.format(self._token)}
r = requests.get(url, headers=headers)
```
and then parse the return for whether a db with that (nice or app) name exists.

**Comment by zikolach** [comment]
Same happens when client ask for list of available applications. My argument is that this feature is not used anywhere (yet) except pyam.
When it is - I am happy to discuss the requirements as currently application name (unique) is used in lots of places from URL segments to request payloads. So I am not sure we want to introduce right now this amount of changes.

**Comment by danielhuppmann** [comment]
> Same happens when client ask for list of available applications.

This is kind-of my point - this implementation creates a lot of overhead even in simple cases... And even getting a list of available applications should be as simple as getting a list from the API, no need to have the mapping in pyam.

Anyway... Since I didn't manage to convince you about an alternative approach, can you clean up this PR such that the tests pass and I will review and merge?

**Comment by zikolach** [comment]
> convince you

It is not about convincing me or not. I see your point to simplify the code, but currently it seems simplifying in one place makes it complicated in another one. Let's postpone this until we have another use for this feature.

> clean up this PR

Tests (at least for me locally) fail because we need to change application config (which may potentially affect users, see related changes in app-configs repository)

File: pyam/iiasa.py (hunk 1 above)
Comment (danielhuppmann) [inline pyam/iiasa.py]: suggest to use `'Duplicate instance alias {}'.format(env)` for consistency

**Comment by danielhuppmann** [comment]
thanks, just checked out this branch and tested it, two issues not in line with expected behaviour:

- [ ] there is a warning `pyam.iiasa - WARNING: Duplicate instance alias prod`
- [ ] the returned list from `pyam.iiasa.Connection().valid_connections` is still the list of application names, not nice names...

**Comment by zikolach** [comment]
>    there is a warning `pyam.iiasa - WARNING: Duplicate instance alias prod`

yes, as there are application instances which configs was not updated to make env (alias) parameter unique. Only changes instance to time is iamc15 to make tests work. Just to confirm - there is nothing to change in pyam in regard of this.

> the returned list from `pyam.iiasa.Connection().valid_connections` is still the list of application names, not nice names...

yes, that's expected behavior. If it is not what is desired please comment how it should work, e.g.:
1. add to the list aliases (will have 2 names per application instance, e.g. IXSE_SR15 and iamc15) simultaneously
2. hide application name when alias is present (IXSE_SR15 will be removed from list, iamc15 will replace it). Possibly break some users scripts (if anyone used).
3. change return type - replace list of strings with e.g. list of tuples or dict or something else (also breaking change).

Please comment on this PR about your desired behavior.

**Comment by danielhuppmann** [comment]
Expected return of `valid_connections` is 2) - use only alias if given (should also help with the other issue, just don't add an `env` parameter), else use app name.

This shouldn't break any code, because connecting via the app name will still work. It just won't be the obvious option when checking via `valid_connections` for available databases.

**Pull request merged**